src/parser/tests/parse_errors/parse_error_statements_tests.rs

```diff
@@ -0,0 +1,976 @@
+                    }),
+                    location: SourceRange::undefined(),
+                },
+                Statement::Assignment {
+                    left: Box::new(ref_to("y")),
+                    right: Box::new(ref_to("x"))
+                }
+            ]
+        )
+    );
+}
+
+#[test]
+fn test_nested_while_with_missing_end_while() {
+    //regress, this used to end in an endless loop
+    let lexer = lex("PROGRAM My_PRG
+            WHILE x = y DO
+                WHILE x = y DO
+                    ;
+                END_WHILE
+                y := x;
+           END_PROGRAM
+    ");
+    let (unit, diagnostics) = parse(lexer).unwrap();
+
+    assert_eq!(
+        diagnostics,
+        vec![
+            Diagnostic::missing_token("[KeywordEndWhile]".into(), (156..167).into()),
+            Diagnostic::unexpected_token_found(
+                "KeywordEndWhile".into(),
+                "'END_PROGRAM'".into(),
+                (156..167).into()
+            ),
+        ]
+    );
+
+    assert_eq!(
+        format!("{:#?}", unit.implementations[0].statements),
+        format!(
+            "{:#?}",
+            vec![Statement::WhileLoopStatement {
+                body: vec![
+                    Statement::WhileLoopStatement {
+                        body: vec![empty_stmt()],
+                        condition: Box::new(Statement::BinaryExpression {
+                            left: Box::new(ref_to("x")),
+                            right: Box::new(ref_to("y")),
+                            operator: crate::ast::Operator::Equal
+                        }),
+                        location: SourceRange::undefined()
+                    },
+                    Statement::Assignment {
+                        left: Box::new(ref_to("y")),
+                        right: Box::new(ref_to("x"))
+                    }
+                ],
+                condition: Box::new(Statement::BinaryExpression {
+                    left: Box::new(ref_to("x")),
+                    right: Box::new(ref_to("y")),
+                    operator: crate::ast::Operator::Equal
+                }),
+                location: SourceRange::undefined(),
+            },]
+        )
+    );
+}
+
+#[test]
+fn test_while_with_missing_do() {
+    //regress, this used to end in an endless loop
+    let lexer = lex("PROGRAM My_PRG
+            WHILE x = y
+                y := x;
+            END_WHILE
+           END_PROGRAM
+    ");
+    let (unit, diagnostics) = parse(lexer).unwrap();
+
+    assert_eq!(
+        diagnostics,
+        vec![Diagnostic::unexpected_token_found(
+            "KeywordDo".into(),
+            "'y := x;'".into(),
+            (55..62).into()
+        ),]
+    );
+
+    assert_eq!(
+        format!("{:#?}", unit.implementations[0].statements),
+        format!(
+            "{:#?}",
+            vec![Statement::WhileLoopStatement {
+                //TODO bad recovery!
```

Comment:
Try to allow the DO keyword here and report if missing (consume_or_report)

Reply:
done